File: 061.Rotate List.java
 /*
    Given a list, rotate the list to the right by k places, where k is non-negative.

    For example:
    Given 1->2->3->4->5->NULL and k = 2,
    return 4->5->1->2->3->NULL.
     */
    /**
     * Definition for singly-linked list.
     * public class ListNode {
     *     int val;
     *     ListNode next;
     *     ListNode(int x) { val = x; }
     * }
     */
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null)
            return head;

        int count = 1;
        ListNode dummy = new ListNode(0);
        ListNode tail = head;

        while (tail.next != null) {
            count++;
            tail = tail.next;
        }

        tail.next = head;

        k %= count;
        if (k != 0) {
            for (int i=0; i<count-k; i++) {
                tail = tail.next;
            }
        }

        dummy.next = tail.next;
        tail.next = null;
        return dummy.next;
    }

File: LeetCode/061.Rotate List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        cnt = 0
        tmp = head
        while tmp:
            tmp = tmp.next
            cnt += 1
        if cnt == 0:
            return head
        k %= cnt
        if k == 0:
            return head

        ptr1 = ptr2 = head
        while k > 0:
            if ptr1 == None:
                ptr1 = head
            ptr1 = ptr1.next
            k -= 1
        while ptr1.next != None:
            ptr1 = ptr1.next
            ptr2 = ptr2.next
        tmp = ptr2.next
        ptr2.next = None
        ptr1.next = head
        head = tmp
        return head
